### ISSUE TITLE
eapply with (name := ...) does not work for implicit arguments

### ISSUE THREAD

**Issue by RalfJung** [coq/coq#6029]
#### Version

8.7.dev

#### Operating system

Linux

#### Description of the problem

The following proof script demonstrates the problem:
```
Lemma lem1 (P Q: Prop) : P -> Q -> P.
Proof. auto. Qed.

Lemma lem2 {P Q: Prop} : P -> Q -> P.
Proof. auto. Qed.

Goal True. Proof.
(* This works *)
eapply lem1 with (Q:=True).
Undo.
(* But this does not *)
eapply lem2 with (Q:=True).
```
As you can see, `lem1` and `lem2` only differ in whether the arguments are implicit and explicit. I would expect `eapply with (name := ...)` not to care about such details, but it does.

This may be related to #4536, because using `@lem2` instead works. So, this is yet another example of maximally implicit arguments having bad side-effects.

**Comment by herbelin** [comment]
@RalfJung: Yes, you exactly diagnosed the problem. The expression `eapply lem2 with (Q:=True)` is the same as `eapply [@lem2 _ _) with (Q:=True)`, so `Q` is already supplied and cannot be found any more.

If enough support can emerge for a change of semantics of `apply c with bindings` so that `c` is restricted to a global reference, then this is probably a change which we could make accepted. After all, most uses of `apply c with bindings` are already of this form.

**Comment by RalfJung** [comment]
It turns out I understand this feature even less than I thought... Coq is rejecting the following, which is baffling:
```
Lemma lem1 (P Q: Prop) (y:nat) (x:nat) : P -> Q -> P.
Proof. auto. Qed.

Goal True. Proof.
eapply lem1 with (Q:=True) (x:=5).
```
It says only `P` and `Q` are bound names. Why is it pretending as if `x` and `y` do not exist?!?

**Comment by ppedrot** [comment]
@RalfJung These variable are not dependent in the type, so they cannot be accessed with the `with` syntax.

**Comment by RalfJung** [comment]
Why the restriction to arguments that occur dependently? That seems rather arbitrary to a user not familiar with the implementation internals.

**Comment by ppedrot** [comment]
The rationale for the `with` semantics is that it should behave "as the user sees it". So if bound variables are renamed for printing, `with` should behave accordingly, and if they're not dependent, they should not be accessible since they have no user-facing name. Do not ask me why we're doing it this way, but this is what is implemented.

**Comment by herbelin** [comment]
> it should behave "as the user sees it"

I would say that this is a remanence of developing proofs without a UI. To know the statement of a lemma, we did not look at where it was stated in a file but at what `Print` or `Show` was giving.

> These variable are not dependent in the type, so they cannot be accessed with the with syntax.

The rule is to use indices to refer to non-dependent variables, counting only non-dependent ones from left-to-right. So `lem1 with (2:=5)` would refer to `x` (the second non-dependent variable). [I know that it is strange to see `2` and `5` with a different role. At some time, I had the suggestion to use ordinals instead, so you would write `lem1 with (2nd:=5)` which I find more intuitive...

> Why the restriction to arguments that occur dependently? 

We could accept to refer to non-dependent but named arguments by name. This requires first to ensure that no names are duplicated in the statement of a global constant, but I believe that there is a rough consensus about that and that @jashug is more or less going in this direction.

**Comment by RalfJung** [comment]
> I would say that this is a remanence of developing proofs without a UI. To know the statement of a lemma, we did not look at where it was stated in a file but at what Print or Show was giving.

I am curious why one wouldn't use `About` to look up the names? That's what I always use anyway and it proved most useful so far. (`Check` does not behave well when there are implicits, it tries to infer them.)

> The rule is to use indices to refer to non-dependent variables, counting only non-dependent ones from left-to-right. So lem1 with (2:=5) would refer to x (the second non-dependent variable). 

Oh, I didn't even know about `<num>:=<val>`, thanks.

**Comment by herbelin** [comment]
> I am curious why one wouldn't use About to look up the names?

Sure! Even if `with` de facto relies on names, the idea that names were a property of constants (modifiable with `Arguments`) is comparatively recent.

> Oh, I didn't even know about `<num>:=<val>`, thanks.

You're welcome.

**Comment by RalfJung** [comment]
> the idea that names were a property of constants (modifiable with Arguments) is comparatively recent.

Ah, that explains many things. But names in constants were always preserved right? With-by-name still basically uses those name, it just refreshes them into the current context.

With https://github.com/coq/coq/pull/12756, that does not happen with `term (name:=val)` any more, so maybe with-by-name should follow suit.
